### PR TITLE
fix: test reliability patches

### PR DESCRIPTION
**Pull request by faddat**
## This PR fixes 

* flakiness in testwriter
* flakiness in testbyzantineprevoteequivocation

@melekes - I'd like to merge this branch, and then open another torture branch and run the tests a few thousand times.  But I think we are in good shape. 

----

## NOTES FOR REVIEWERS

Some of the fixes here involved small changes to the tolerances of tests in situations where the code may behave in a slightly variable way without malfunctioning.  I bet that is the key item for review. 

This needs to be broken into pieces.

~~I consider this branch to be mergable now~~, and I suppose what I'll do is make another one where the other tests I identified as having issues are run 100 times.  That was totally key to fixing `TestTransportMultiplexAcceptNonBlocking` because it got me the data I needed to understand the problem.

---

#### PR checklist

- [x] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

**Comment by melekes** [comment]
`TestByzantinePrevoteEquivocation` and `TestTransportMultiplexAcceptNonBlocking` are still failing.

```
--- FAIL: TestByzantinePrevoteEquivocation (30.08s)
    byzantine_test.go:292: Evidence from each validator: [<nil> <nil> <nil> <nil>]
    byzantine_test.go:293: Timed out waiting for validators to commit evidence

--- FAIL: TestTransportMultiplexAcceptNonBlocking (0.04s)
    transport_test.go:360: have {{9 11 0} dc28ea6b630e74a67256457f901d9d183da919d7 127.0.0.1:38953 testing 1.2.3-rc0-deadbeef 01 slow_peer {on 127.0.0.1:44371}}, want {{9 11 0} 6[123](https://github.com/cometbft/cometbft/actions/runs/7390672952/job/20106075991?pr=1954#step:5:124)f7e124d306734abc6a4d438a5a734f322dcc 127.0.0.1:37183 testing 1.2.3-rc0-deadbeef 01 fastnode {on 127.0.0.1:35069}}
```

**Comment by faddat** [comment]
Working on it sir :D

**Comment by faddat** [comment]
@melekes probably obvious, but I'm using this branch as a testing ground for all of the tests at once since we need to run them many times to call em flaky or not.

**Comment by faddat** [comment]
PS: TestTransportMultiplexAcceptNonBlocking is the hardest one imo.

It is testing that a slow peer doesn't block a fast peer.

**Comment by faddat** [comment]
#1948 is still failing at 35 seconds.  Do you think we should check out the evidence submission code, or bump to 40 seconds?

@melekes

**Comment by faddat** [comment]
I'm going to bump to 40.... but i think we could have a code or test issue on that one

**Comment by faddat** [comment]
oh darn it and I am comitting to the wrong branch.  But making actual progress, so at least the commits will be useful...

**Comment by faddat** [comment]
testbyzantineprevoteequivocation is very challenging.

**Comment by faddat** [comment]
tfw a seven year old go lib solves a problem:

https://github.com/phayes/freeport

**Comment by faddat** [comment]
When this passes, I'll cherry pick changes or check out files into the individual PRs -- or if preferred, we can merge this as "test reliability patches"

File: rpc/jsonrpc/jsonrpc_test.go [hunk 1]
```diff
@@ -21,14 +22,13 @@ import (
 	server "github.com/cometbft/cometbft/rpc/jsonrpc/server"
 	types "github.com/cometbft/cometbft/rpc/jsonrpc/types"
 	"github.com/go-kit/log/term"
+	"github.com/phayes/freeport"
```
Comment (faddat) [inline rpc/jsonrpc/jsonrpc_test.go:25]: note to self: let's eliminate this dependency

**Comment by faddat** [comment]
Jobs are currently just staying queued, so I'll take a break from this for today :)

See ya tomorrow :)

**Comment by faddat** [comment]
@melekes - for merge conflict purposes, I think this one should be merged before #1906 if that is alright with ya.

**Comment by faddat** [comment]
@melekes merged your branch here to see if it would stop some tests from failing.

fairly sure that'll be the case.

Goal: make this branch green after 40-1000x runs, then either merge the whole thing or break it into individual PR's at your preference.

**Comment by faddat** [comment]
the "try delaying to not block ports" ensures we test on the default ports.  It might be better to randomize as we do elsewhere though.

**Comment by faddat** [comment]
This branch is no longer mergeable and I should break it into individual changes.

**Pull request closed without merging**